### PR TITLE
use snapshots for SNS tests

### PR DESCRIPTION
**Pull request by bentsku**
This PR adds the use of snapshots for every `aws_validated` test in the SNS test suite, following #6586
We skip some paths for now to be fixed in a later iteration. It is mostly fields not added (or added when it should not be) by Moto.
Also, it seems AWS doesn't add message signature for Fifo topics, so we will need to remove those fields. 
I added some exception messages fixes as well.

It contains a hacky fix that will be added in #6598 and will be removed then in order to guarantee the order of the fields.

### CODE REVIEW

File: localstack/testing/snapshots/transformer_utility.py
Comment (thrau) [inline localstack/testing/snapshots/transformer_utility.py:171]: I had a similar issue in #6603. It seems related to the the fact that the `SenderId` of an SQS message can sometimes be the root account ID, and sometimes the access key id, see. https://stackoverflow.com/a/28603793/804840
This then messes up the comparisons.

Any idea what a good pattern is for this @steffyP @dominikschubert ?

File: localstack/testing/snapshots/transformer_utility.py
Comment (thrau) [inline localstack/testing/snapshots/transformer_utility.py:186]: should we not maybe leave this in per default? or are you having issues with bodies containing data that is dynamically generated by the tests? (like passing ARNs through a message body).

File: localstack/testing/snapshots/transformer_utility.py
Comment (thrau) [inline localstack/testing/snapshots/transformer_utility.py:185]: should we maybe just use the default `value_replacement`?

**Review by thrau — APPROVED** [review]
Looks fantastic, snapshots allover :camera_flash: :sparkles: 

I'll let @dominikschubert and @steffyP comment/decide on the changes in the transformer utils.

File: localstack/testing/snapshots/transformer_utility.py
Comment (steffyP) [reply localstack/testing/snapshots/transformer_utility.py:171]: @thrau: Interesting :) I think a key-value replacement should be fine in that case. We could also include a regex-check to make sure it matches one of the two valid patterns. but seems like a overhead in that case.

Only thing that feels a bit hacky here: if we replace it with 1s, one could assume that the region-regex-replacement was used here - which is not the case. @bentsku is there a particular reason why you have chosen this value?

File: localstack/testing/snapshots/transformer_utility.py
Comment (steffyP) [reply localstack/testing/snapshots/transformer_utility.py:186]: +1 
If the md5 hash matches in most cases, we should not always exclude it. Is this the case for SNS?
Ideally we would only exclude md5-hashes for individual test cases with a note why it has to be replaced (e.g. includes timestamp).

File: localstack/testing/snapshots/transformer_utility.py
Comment (steffyP) [reply localstack/testing/snapshots/transformer_utility.py:185]: with the reference_replacement set to false, the default value would be `signature`. I think it's fine to use  `<signature>` here as it indicates at the first sight that the value was replaced.

File: localstack/testing/snapshots/transformer_utility.py
Comment (steffyP) [inline localstack/testing/snapshots/transformer_utility.py:170]: @bentsku: is there a specific reason why you decided to replace the senderId with 1s? IMO it can be a bit misleading, as we also have the regex-replacement for the account-id, and on first sight it will look like the regex-replacement is responsible for this replacement. 

Personally, I would prefer if we could use distinct replacements values for transformers.

File: localstack/testing/snapshots/transformer_utility.py
Comment (bentsku) [reply localstack/testing/snapshots/transformer_utility.py:186]: The MD5 hash almost never matches: the body of the SNS message always contains a timestamp.
It only matches when `RawDelivery` is set to True, as the body is the string message passed. I had to manually add the MD5 transformer to almost every test. I thought the MD5 hash could just be checked with a unit test.
Otherwise, I had an over engineered way to do, which was to delete the transformer from the `sns_api`.
What do you think we should do in this case?

**Review by steffyP — APPROVED** [review]
Nice to see so many snapshot tests 😄 📸 😍 

Only one thing that bothers me a bit: the `SenderId` replacement somehow shadows the regex accountid replacement. Maybe we can improve this?

Regarding the [snapshot-PR](https://github.com/localstack/localstack/pull/6598) you are depending on: I am waiting for a review from @dominikschubert and his input :)

File: localstack/testing/snapshots/transformer_utility.py
Comment (bentsku) [reply localstack/testing/snapshots/transformer_utility.py:185]: Yes that was my issue, the fact that the `<[...]>` were missing was troubling me, so I added it manually!

File: localstack/testing/snapshots/transformer_utility.py
Comment (bentsku) [reply localstack/testing/snapshots/transformer_utility.py:170]: I am going to test it again, but I believe the issue was that locally, the `SenderId` was replaced by 1s, and not in the tests against AWS. I'll check again, could be priorities, I need to check what the response was before transformation.

File: localstack/testing/snapshots/transformer_utility.py
Comment (steffyP) [reply localstack/testing/snapshots/transformer_utility.py:186]: alright :) i think in that case, having unittests should be fine. 
alternatively we could also create another api-helper, which excludes the md5-transformer and use this one for the `RawDelivery` messages

Comment (bentsku) [reply localstack/testing/snapshots/transformer_utility.py:186]: Yes, it could work as well. The issue also that I didn't think about is that in some tests, we have both `RawDelivery` and regular in the same test. So I have to ignore the hash. Maybe I will manually check the MD5 of raw messages in the integration tests? As it's AWS validated, it will check the behaviour when we regenerate the snapshots

Comment (steffyP) [reply localstack/testing/snapshots/transformer_utility.py:186]: I see :) yes, sounds good. but I think it would also be fine to cover the md5-checksum in unittests as you suggested before.

File: localstack/testing/snapshots/transformer_utility.py
Comment (bentsku) [reply localstack/testing/snapshots/transformer_utility.py]: This is now changed and simply use `reference_replacement=False` instead, as it is enough to not mess with ARN anymore

File: localstack/testing/snapshots/transformer_utility.py
Comment (bentsku) [reply localstack/testing/snapshots/transformer_utility.py:185]: The md5 is part of SQS, and is covered inside the integration tests there, there shouldn't be any regression on that side.

**Comment by bentsku** [comment]
Thanks a lot for your reviews, I updated the PR with your comments.
We should not merge until https://github.com/localstack/localstack/pull/6598 is merged, I'm worried the ordering could break the tests as I am using it only partially in my branch. I will rebase on master once it is done and regenerate the tests that fails, if they do.
Thanks again!

**Review by steffyP — APPROVED** [review]
LGTM! thanks for cleaning up the tests and for your help with improving the snapshot transformers 🚀 👍

**Pull request merged**
